Issue 41864: DataRegion column analytics updates to lazy load Ext4 and vis dependencies

cnathe opened the pull request:
#### Rationale
Issue [41864](https://www.labkey.org/home/Developer/issues/issues-details.view?issueId=41864): Remove dataregion dependency on ExtJS (SummaryStatisticsAnalyticsProvider)

#### Changes
* moves LABKEY.ColumnSummaryStatistics.showDialogFromDataRegion from ColumnSummaryStatistics.js to DataRegion.js
* remove ColumnSummaryStatistics dependency and vis dependency from AnalyticsProvider classes
* fix DataRegion update to column header menu item disabled state on add/remove column visualization
* wrap ColumnVisualizationAnalytics.js functions in calls to LABKEY.requiresVisualization()

cnathe committed:
Issue 41864: DataRegion column analytics updates to lazy load Ext4 and vis dependencies
- moves LABKEY.ColumnSummaryStatistics.showDialogFromDataRegion from ColumnSummaryStatistics.js to DataRegion.js
- remove ColumnSummaryStatistics dependency and vis dependency from AnalyticsProvider classes
- fix DataRegion update to column header menu item disabled state on add/remove column visualization
- wrap ColumnVisualizationAnalytics.js functions in calls to LABKEY.requiresVisualization()

labkey-nicka commented on api/webapp/clientapi/dom/DataRegion.js:
```diff
@@ -2831,6 +2827,16 @@ if (!LABKEY.DataRegions) {
     };
 
+    var _updateAnalyticsProviderMenuItem = function(elementId, disable) {
+        var el = $('#' + elementId.replace(/:/g, '\\:'));
```
nit: This line could use a comment. Not exactly clear why this replacement is being done.

cnathe replied:
done

labkey-nicka commented on api/webapp/clientapi/dom/DataRegion.js:
```diff
@@ -4742,4 +4748,44 @@ LABKEY.AggregateTypes = {
+                    region.getColumnAnalyticsProviders(regionViewName, colFieldKey, function(colSummaryStats) {
+                        Ext4.create('LABKEY.ext4.ColumnSummaryStatisticsDialog', {
+                            queryConfig: region.getQueryConfig(),
+                            filterArray: LABKEY.Filter.getFiltersFromUrl(region.selectAllURL, 'query'), //Issue 26594
```
Shouldn't this be using the data region's `dataRegionName`?

```js
filterArray: LABKEY.Filter.getFiltersFromUrl(region.selectAllURL, region.dataRegionName),
```

I realize this code is just copied across so take it or leave it.

cnathe replied:
Yeah, I just copied this code over and wasn't planning to change it at all, but I can take a look at this suggestion.

cnathe replied:
Turns out that the `region.selectAllURL` which is being used in this call to `LABKEY.Filter.getFiltersFromUrl()` does actually use the constant `query` as the url param prefix instead of the `dataRegionName`.

labkey-nicka replied:
Eesh, that sounds like a bug...

labkey-nicka commented on api/webapp/clientapi/dom/DataRegion.js:
```diff
@@ -4742,4 +4748,44 @@ LABKEY.AggregateTypes = {
     // TODO how to allow premium module additions to aggregate types?
+};
+
+LABKEY.ColumnSummaryStatistics = new function () {
```
Since this is now defined in `DataRegion.js` it might be better to just incorporate this into `DataRegon` much like customize view. Then the callers can call `LABKEY.DataRegion[regionName].showSummaryStatisticsDialog(colFieldKey)` or some such.

This would remove the need for this extra `LABKEY.ColumnSummaryStatistics` namespace and feel more "native" to a data region.

cnathe replied:
Good call, I'll make that change.

cnathe replied:
Done

labkey-nicka commented on api/src/org/labkey/api/view/PopupMenuView.java:
```diff
@@ -157,6 +157,8 @@ else if ("-".equals(child.getText()))
     protected static void renderTreeItem(NavTree item, String cls, Writer out) throws IOException
     {
         out.write("<li");
+        if (item.getId() != null)
```
Hmm, I don't think this is a good idea. This id is not guaranteed to be unique or compliant with the DOM id accepted syntax. Additionally, it ends up being generated for a lot more elements than just the one we're looking to target with this change.

That said, I presume you got this code to invoke `_updateAnalyticsProviderMenuItem` (I couldn't quite find the scenario). Could we instead load Ext4 at the time this is called as well and then still do the component query?

cnathe replied:
Yeah, I see what you are saying about the DOM id. We could revert back to the Ext4 component query, but from what I could tell that was actually broken as well since it was trying to do an elementId query and from what I could tell that wasn't applicable any more (i.e. these aren't Ext4 menu anymore). I think we could instead do a jQuery selector without having to add this id. I'll take another look.

cnathe replied:
Done

cnathe committed:
Merge branch 'develop' into fb_issue41864_columnAnalytics

cnathe committed:
LABKEY.ColumnSummaryStatistics.showDialogFromDataRegion -> LABKEY.DataRegion[regionName].showColumnStatisticsDialog

cnathe committed:
Revert change to PopupMenuView

cnathe committed:
Update jQuery selector for column analytics provider menu item to work without specific id
- instead use selector for column-name attr and then child by analytics provider name

labkey-nicka reviewed (approved):
Awesome. I think this is a good change.
